design/open/staging-compaction.md

```diff
@@ -0,0 +1,172 @@
+# Proposal: Staging Compaction
+
+## Problem Description
+
+lakeFS staging area is built on top of the [KV store](../accepted/metadata_kv/index.md).
+In short, the structure for the `Branch` entity in the kv-store is as follows:
+```go
+type Branch struct {
+	CommitID     CommitID
+	StagingToken StagingToken
+	// SealedTokens - Staging tokens are appended to the front, this allows building the diff iterator easily
+	SealedTokens []StagingToken
+}
+```
+
+Uncommitted entries are read from the staging token first, then from the 
+sealed tokens by order. Writes are performed on the staging token.
+
+The KV design has proven to meet lakeFS requirements for its consistency 
+guarantees. For most use-cases of uncommitted areas in lakeFS, it does that 
+efficiently. However, there are some cases where it falls short, for example 
+this [issue](https://github.com/treeverse/lakeFS/issues/2092).  
+There might be other cases where the structure of the `Branch` entity impacts 
+the performance of reading from the staging area, for example when the 
+number of sealed tokens is large (N)[^1] and reading a missing entry requires 
+reading from all N sealed tokens. 
+
+## Goals
```

Comment:
I don't see any reference to visibility into this internal feature. 
To me this is parallel to having a GC, I wanna know if it worked, when it runs, how much time did it take etc. 
If you agree can you add a small section about the visibility requirements?

Reply:
As a user, why do you want to know if compaction worked? And how effective was it? You only care about it when performance degrades and you need to drill down further. I guess most users will never hear of that.
GC is different. It's an external process I need to manage by myself (or pay someone to do it) and it results in storage costs.
We should have metrics for it just like with any other flow in Graveler.

Reply:
First, sorry for not being clear - I meant GC in JVM not lakeFS. 
And this is super internal but, we must have some visibility into the performance of compaction isolated from the rest. 
The reason is because the sensor would evolve over time into different directions with heuristics / tradeoffs / configuration. 
Having visibility is important. 

I think that this should be mentioned as part of the design, otherwise this can easily be missed and "shipped later" or done with an after thought.

Reply:
I think prometheus metrics are more than enough for the sophisticated lakeFS user to tweak what needs to be tweaked. What new metrics would you like me to call out in this design?

Reply:
compaction successful runs number, compaction failures number or total compaction runs.

Reply:
So a single event `compaction_run {success:bool}`